### ISSUE TITLE
Bug: [no-misused-promises] Flags an async executor passed to Promise

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#7405]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsI2AZQKZSYAmAFAJQwDeAUDNgO7CrIwR70wAKATugLaoQeEtzwh0AGwBueGAF4AfDCHIAKqj550CZCLGSZAGhgBGAAwWyZANzUAvtWpsOPfoOGhIsPeOl4KilS0DEwsuATE5LZ0or4yJABEABZ4EhLoCTb2ZEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=&tokens=false

### Repro Code

```typescript
async function aSecond() {
  await new Promise(resolve => setTimeout(resolve, 1000));
}

new Promise(async (resolve) => {
  await aSecond();
  resolve("hello");
})
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-misused-promises": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I'm under the impression it's reasonable to pass an async function to a `Promise`. So this shouldn't report a complaint, right?

### Actual Result

```plaintext
Promise returned in function argument where a void return was expected. 5:13 - 8:2
```

### Additional Info

Looking at `lib.es2015.promise.d.ts`, `PromiseConstructor`'s `new` is typed as the `executor` returning `void`, not `void | PromiseLike`:

```ts
new <T>(
    executor: (
        resolve: (value: T | PromiseLike<T>) => void,
        reject: (reason?: any) => void
    ) => void
): Promise<T>;
```

Perhaps this is an issue in the TypeScript types?

**Comment by Josh-Cena** [comment]
> I'm under the impression it's reasonable to pass an async function to a `Promise`.

Why would you do that, though? You could be doing this:

```js
aSecond().then(() => new Promise((resolve) => {
  resolve("hello");
});
```

There's even a core rule to forbid async executors, because they don't make much sense: https://eslint.org/docs/latest/rules/no-async-promise-executo

**Comment by bradzacher** [comment]
Yeah @voxpelli has it right - this is working as intended because an async promise executor will create a "floating" promise with no handler and if it throws then it'll be an unhandled promise rejection.

@Josh-Cena's suggestion is a better solution - relying on promise chaining will ensure the promise is correctly handled.

I always prefer `async` though as chaining is fraught with footguns.
If for some reason you can't use `await` in the context you can also sub-context via an IIFE:
```ts
return (async () => {
  await aSecond();
  return 'hello';
})();
```
